keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1073 @@
+7. Extensible: should work with custom policy implementations without whole-sale replacement
+    - Enable custom policies keyed off of RuntimeClassNames
+
+Nice to have:
+
+1. Exceptions or policy bindings by requesting user
+2. (fuzzy) Windows support in the initial release
+3. Admission controller is enabled by default in beta or GA phase
+4. Provide an easy migration path from PodSecurityPolicy
+5. Enforcement on pod-controller resources (i.e. things embedding a PodTemplate)
+
+### Non-Goals
+
+1. Provide a configurable policy mechanism that meets the needs of all use-cases
+2. Limit privilege escalation and other attacks beyond the pod to host boundary
+   (e.g. policies on services, secrets, etc.)
+3. Feature parity with PodSecurityPolicy. In particular, support for providing default values or any
+   other mutations will not be included.
+
+## Proposal
+
+The three profile levels (privileged, baseline, restricted) of the [Pod Security Standards] will
+be hardcoded into the new admission plugin. Changes to the standards will be tied to the Kubernetes
+version that they were introduced in.
+
+### API
+
+Policy application is controlled based on labels on the namespace. The following labels are supported:
+```
+pod-security.kubernetes.io/enforce: <policy level>
+pod-security.kubernetes.io/enforce-version: <policy version>
+pod-security.kubernetes.io/audit: <policy level>
+pod-security.kubernetes.io/audit-version: <policy version>
+pod-security.kubernetes.io/warn: <policy level>
+pod-security.kubernetes.io/warn-version: <policy version>
+```
+
+**Enforce:** Pods meeting the requirements of the enforced level are allowed. Violations are rejected
+in admission.
+
+**Audit:** Pods and [templated pods] meeting the requirements of the audit policy level are ignored.
+Violations are recorded in a `pod-security.kubernetes.io/audit-violations: <violation>` [audit
+annotation](#audit-annotations) on the audit event for the request. Audit annotations will **not**
+be applied to the pod objects themselves, as doing so would violate the non-mutating requirement.
+
+**Warn:** Pods and [templated pods] meeting the requirements of the warn level are ignored.
+Violations are returned in a user-facing warning message. Warn & audit modes are independent; if the
+functionality of both is desired, then both labels must be set.
+
+[templated pods]: #podtemplate-resources
+
+There are several reasons for controlling the policy directly through namespace labels, rather than
+through a separate object:
+
+- Using labels enables various workflows around policy management through kubectl, for example
+  issuing queries like `kubectl get namespaces -l
+  pod-security.kubernetes.io/enforce-version!=v1.22` to find namespaces where the enforcing
+  policy isn't pinned to the most recent version.
+- Keeping the options on namespaces allows atomic create-and-set-policy, as opposed to creating a
+  namespace and then creating a second object inside the namespace.
```

Comment:
> > If we have a separate cluster scoped object called PodSecurity where the name of object matches the name of the namespace, one interesting property is that authorization to write said PodSecurity is distinct from write on namespace.
> 
> I feel like we discussed this before, but I can't remember if there were any concerns aside from losing out on the label-selector UX. If we went this route, we'd probably want to add a custom kubectl command for it, but it would need to be fairly complicated to cover all the use cases we'd get for free with the label selector.

Note that this mostly a thought exercise of what we lose by using label selectors instead of a distinct object.

Reply:
> > So should we define extra authorization (SAR checks with virtual verbs similar to PSP's use) checks needed to set these labels and enforce them in admission?
> 
> I think there is a use case for generic label policy, and I'd be interested in a proposal for it, but IMO we shouldn't implement something special for pod security.

I do not know if I buy the "let us wait for the generic label policy" approach, especially since we want the pod security stuff to mostly be static.  Having a one-off approach for pod security that encodes some well known permissions for the pod security labels could provide significant value and safety to this feature, especially in environments where users are allowed to provision namespaces.